SONARIAC-361 Rule 6440: IAM resources

stanislav-jordanov-sonarsource opened the pull request.

nils-werner-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/gcp/IamResourcesCheckPart.java:
```diff
@@ -0,0 +1,157 @@
+      .ifPresent(attr -> ctx.reportIssue(attr, String.format(MESSAGE, ((TextTree) attr.value()).value())));
+  }
+
+  private static boolean isaPrivilegedRole(String roleName) {
+    // Bear in mind PRIVILEGED_ROLES are all CAPS:
+    roleName = roleName.toUpperCase();
+    for (String privilegedRole : PRIVILEGED_ROLES) {
+      if (roleName.contains(privilegedRole)) {
+        return true;
+      }
+    }
+    return false;
+  }
```
Fine for me, but instead of a loop calling 4 times contains and toUpperCase on the string. I would simply create a small predicate based on a pattern and use Pattern.CASE_INSENSITIVE as a flag.

nils-werner-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/gcp/IamResourcesCheckPart.java:
```diff
@@ -0,0 +1,157 @@
+    return false;
+  }
+
+  String[] IAM_RESOURCE_NAMES = {
```
I like the idea to have such a huge list at the end of the file even when this is a constant. However, it should be private final strict. WDYT?

nils-werner-sonarsource commented on iac-extensions/terraform/src/test/java/org/sonar/iac/terraform/checks/gcp/AdministrativeRolesOnWorkloadResourcesCheckTest.java:
```diff
@@ -0,0 +1,31 @@
+class AdministrativeRolesOnWorkloadResourcesCheckTest {
+
+  @Test
+  void test() {
```
```suggestion
  void role_assignment() {
```

nils-werner-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/gcp/IamResourcesCheckPart.java:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  private static void checkIamResources(CheckContext ctx, BlockTree resource) {
+    PropertyUtils.get(resource, "role", AttributeTree.class)
+      .filter(attr -> TextUtils.matchesValue(attr.value(), IamResourcesCheckPart::isaPrivilegedRole).isTrue())
+      .ifPresent(attr -> ctx.reportIssue(attr, String.format(MESSAGE, ((TextTree) attr.value()).value())));
```
Why did you use the old approach? I would guess using the new style with `resource.attribute` looks smarter.

stanislav-jordanov-sonarsource replied:
made it even smarterER than requested :)

nils-werner-sonarsource commented on iac-extensions/terraform/src/test/resources/checks/GCP/AdministrativeRolesOnWorkloadResourcesCheck/role_assignment.tf:
```diff
@@ -0,0 +1,62 @@
+}
+
+resource "google_apigee_environment_iam_member" "member" {
+  org_id = google_apigee_environment.apigee_environment.org_id
+  env_id = google_apigee_environment.apigee_environment.name
+  role = "roles/apigee.admin" # NonCompliant
+  member = "user:jane@example.com"
```
```suggestion
  role = "roles/apigee.admin" # Noncompliant
```
1. I guess you have to write `Noncompliant` instead of `NonCompliant` but no bet
2. You can get rid of the overhead because it is not necessary for the check. Having it on the first test case should be enough.

stanislav-jordanov-sonarsource replied:
`# NonCompliant` works too (tested) but okay - complying

stanislav-jordanov-sonarsource commented:
1. Renamed and reordered the 6 previously existing ResourceVisitor.Attribute.reportIfXyz(..) methods. I claim this aids code readability.
2. Added 2 additional methods: `reportIfValueMatches(Pattern ...) & reportIfValueContains(Pattern ...)` (added in a previous commit)
3. Replaced `RoleScopeHelper.matchPredicate(String)` with 4 more fine-grained Predicate<String> creator methods
4. Added `RoleScopeHelper.treePredicate(Predicate<String>)` to build a Predicate<Tree> from a Predicate<String>
5. Optimized away some hotspots that were compiling the same regex Pattern(s) on any subsequent method invocation.

nils-werner-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/ResourceVisitor.java:
```diff
@@ -162,22 +163,33 @@ public Attribute reportIfFalse(String message, SecondaryLocation... secondaries)
     }
 
-    public Attribute reportIfValueDoesNotMatch(Predicate<ExpressionTree> expectedPredicate, String message, SecondaryLocation... secondaries) {
+    public Attribute reportIfNot(Predicate<ExpressionTree> expectedPredicate, String message, SecondaryLocation... secondaries) {
       // designed to be extended but noop in standard case
       return this;
     }
```
This method is only used in test cases. So it seems that there is no real use case so far.

stanislav-jordanov-sonarsource replied:
I've just renamed
`reportIfValueDoesNotMatch(Predicate<ExpressionTree> ...)` to
`reportIfNot(Predicate<ExpressionTree> ...)`
I did not plan to remove existing methods.
Or are do you mean some other method?

stanislav-jordanov-sonarsource replied:
Besides, if this one is missing, future users (i.e. we) will be encouraged to use
`reportIf(<predicate>.negate())`
which is both less readable and creates a new wrapper predicate on each invocation.
Due to this I suggest we keep it.

nils-werner-sonarsource reviewed (approved):
👍 Implementation of the rule looks good
👍 Adding a lot of new helper methods to RoleScope helper is out of scope
👍 Adding some valuable methods to ResourceVisitor is valid
👎 Adding unused helper methods is out of scope and blows up the PR

🙃 During reviewing the PR I was a bit shocked by the number of report methods we have. I have an idea to clean up ResourceVisitor. To not block your work, please address the suggestions and merge your PR to the feature branch.

nils-werner-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/CertificateBasedAuthenticationCheck.java:
```diff
@@ -21,12 +21,19 @@
 
 import org.sonar.check.Rule;
 import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.tree.Tree;
```
Unsued import.

nils-werner-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/CertificateBasedAuthenticationCheck.java:
```diff
@@ -21,12 +21,19 @@
+
+import static java.util.regex.Pattern.CASE_INSENSITIVE;
+import static org.sonar.iac.terraform.checks.azure.helper.RoleScopeHelper.exactMatchStringPredicate;
+import static org.sonar.iac.terraform.checks.azure.helper.RoleScopeHelper.treePredicate;
```
Unsued import.

stanislav-jordanov-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java:
```diff
@@ -45,8 +47,26 @@ private RoleScopeHelper() {
+    return exactMatchStringPredicate(regex, 0);
+  }
+
+  public static Predicate<String> exactMatchStringPredicate(String regex, int flags) {
+    final Pattern compiledPattern = Pattern.compile(regex, flags);
+    return s -> compiledPattern.matcher(s).matches();
+  }
```
Consider `exactMatchStringPredicate` as a factory method (akin to c-tor) that creates exactly what its name reads:
Given a String regex literal, this method will create a Predicate<String> (hence the 'StringPredicate' suffix) that will return `true` on the tested String only iff the regex exactly matches the whole tested String.

In a similar fashion `containsMatchStringPredicate` will create a String Predicate that will return `true` iff the regex is able to match a substring of the tested String. 'contains' was chosen due to the semantics of `java.lang.String.contains(CharSequence substring)` method.

stanislav-jordanov-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/azure/helper/RoleScopeHelper.java:
```diff
@@ -45,8 +47,26 @@ private RoleScopeHelper() {
 
-  public static Predicate<String> matchPredicate(String pattern) {
-    return s -> Pattern.matches(pattern, s);
+  public static Predicate<String> exactMatchStringPredicate(String regex) {
+    return exactMatchStringPredicate(regex, 0);
+  }
+
+  public static Predicate<String> exactMatchStringPredicate(String regex, int flags) {
+    final Pattern compiledPattern = Pattern.compile(regex, flags);
+    return s -> compiledPattern.matcher(s).matches();
+  }
+
+  public static Predicate<String> containsMatchStringPredicate(String regex) {
+    return containsMatchStringPredicate(regex, 0);
+  }
+
+  public static Predicate<String> containsMatchStringPredicate(String regex, int flags) {
+    final Pattern compiledPattern = Pattern.compile(regex, flags);
+    return s -> compiledPattern.matcher(s).find();
+  }
+
+  public static <T extends Tree> Predicate<T> treePredicate(Predicate<String> stringPredicate) {
+    return tree -> TextUtils.matchesValue(tree, stringPredicate).isTrue();
```
How about moving them to ResourceVisotor ?

stanislav-jordanov-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/ResourceVisitor.java:
```diff
@@ -162,22 +163,33 @@ public Attribute reportIfFalse(String message, SecondaryLocation... secondaries)
     }
 
-    public Attribute reportIfValueMatches(Predicate<ExpressionTree> expectedPredicate, String message, SecondaryLocation... secondaries) {
+    public Attribute reportIfValueMatches(Pattern expectedValuePattern, String message, SecondaryLocation... secondaries) {
+      // designed to be extended but noop in standard case
+      return this;
+    }
```
This is used now (not only in tests)

stanislav-jordanov-sonarsource commented on iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/gcp/AdministrativeRolesOnWorkloadResourcesCheck.java:
```diff
@@ -0,0 +1,32 @@
+  public void initialize(InitContext init) {
+    new IamResourcesCheckPart().initialize(init);
+  }
+}
```
WIGOH!

stanislav-jordanov-sonarsource commented on iac-extensions/terraform/src/test/resources/checks/GCP/AdministrativeRolesOnWorkloadResourcesCheck/role_assignment.tf:
```diff
@@ -0,0 +1,38 @@
+resource "google_apigee_environment_iam_binding" "binding" {
+  org_id = google_apigee_environment.apigee_environment.org_id
+  env_id = google_apigee_environment.apigee_environment.name
+  role  = "roles/ml.jobOwner" # Noncompliant
```
WIGOH!
